### PR TITLE
Reader: Comment Deep Linking (Permalinks)

### PR DESCRIPTION
**Pull request by samouri**
Currently the Reader doesn't support linking directly to a comment. 
This in particular is a blocker for https://github.com/Automattic/wp-calypso/pull/14396, as well as other improvements we have planned down the line.

In order to get us all the way to being capable of linking directly to comments there are a few steps:
1.  [x] enable bidirectional paging from any comment to earlier/later comments
2. [x] enable loading a specific comment directly.  https://github.com/Automattic/wp-calypso/pull/16006
3. [x] make full-post pay attention to anchors in the url and start from the specified comment

To Test:
- [x] comments in My Sites still render properly.  
- [x] comments in Reader still render properly when not deep linking
- [x] going to a Reader full-post url that ends in #comment-${id} renders / scrolls you to that specific comment with the ability to paginate if necessary in either direction.  I recommend testing multiple posts with different number of comments: 100s, < 10, and 1.
- [x] works with invalid commentIds (doesnt break calypso)
- [x] works with valid commentId but for the wrong post.

### CODE REVIEW

File: client/blocks/comments/post-comment-list.jsx [hunk 1]
```diff
@@ -210,110 +236,115 @@ class PostCommentList extends React.Component {
...
-					</SegmentedControl>
-				}
+					</SegmentedControl> }
 				{ this.renderCommentsList( displayedComments ) }
+				{ showViewLater &&
+					<span
+						className="comments__view-earlier"
+						onClick={ this.viewEarlierCommentsHandler }
+					>
+						{ translate( 'View later comments (Showing %(shown)d of %(total)d)', {
```
Comment (a8ci18n) [inline client/blocks/comments/post-comment-list.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'View earlier comments (Showing %(shown)d of %(total)d)' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by samouri** [comment]
Making a note here that it looks like the fullpostview is rerendering more than it should be

File: client/blocks/comments/post-comment-list.jsx [hunk 2]
```diff
@@ -56,25 +69,65 @@ class PostCommentList extends React.Component {
...
 	}
 
 	componentWillReceiveProps( nextProps ) {
 		const nextSiteId = get( nextProps, 'post.site_ID' );
 		const nextPostId = get( nextProps, 'post.ID' );
 		const nextCommentsFilter = get( nextProps, 'commentsFilter' );
+		const nextStartingCommentId = get( nextProps, 'startingCommentId' );
+		const { hasReceivedBefore, hasReceivedAfter } = get( nextProps, 'commentsFetchingStatus' );
 
 		if (
```
Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: its messy as hell, but this creates the 2-step comment loading behavior (first comment + page before/after).  I can probably clear it up a bit w/ good function names

**Comment by samouri** [comment]
I have a few tests to update :(

File: client/blocks/comments/post-comment-list.jsx [hunk 3]
```diff
@@ -23,28 +19,45 @@ import CommentCount from './comment-count';
 import SegmentedControl from 'components/segmented-control';
 import SegmentedControlItem from 'components/segmented-control/item';
 
+// eslint-disable
```
Comment (bluefuton) [inline client/blocks/comments/post-comment-list.jsx]: Which eslint rule does this trip up? `no-helpful-comment-blocks`?!

Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: `valid-jsdoc` in about a hundred ways heh.
Found a better fix.  just by creating a newline separation it won't treat it as describing the class

**Comment by samouri** [comment]
rebased!

File: client/blocks/comments/post-comment-list.jsx [hunk 4]
```diff
@@ -55,26 +67,76 @@ class PostCommentList extends React.Component {
...
+		}
 	}
 
 	componentWillReceiveProps( nextProps ) {
 		const nextSiteId = get( nextProps, 'post.site_ID' );
 		const nextPostId = get( nextProps, 'post.ID' );
 		const nextCommentsFilter = get( nextProps, 'commentsFilter' );
 
-		if (
+		if ( this.shouldFetchInitialComment( nextProps ) ) {
```
Comment (blowery) [inline client/blocks/comments/post-comment-list.jsx]: we probably need to reset `this.hasScrolledToComment` somewhere in here when the site ID / post ID changes?

Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: hmm. good callout.  I'm not sure that would ever happen in practice yet (zooming from fullpost permalink to fullpost permalink). But we should handle it

Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: > use case would be

oh whelp.  good idea. I'll throw it in the if statement.

**Comment by samouri** [comment]
Turns out if you give it a commentId that doesn't exist for the post, all hell breaks loose.
I'll need to fix that 🤔

**Comment by samouri** [comment]
I fixed two edge cases: 

1. what if the comment doesn't exist or for whatever reason the network requests 404 (permissions etc).  The component needs to give up and request for comments the regular way.
2. what if the comment _does_ exist but for a different post on the site! component needs to do the same thing + give up + request comments the regular way.

Both of these are relatively tricky + don't easily fit into the data-layer paradigm as far as I can tell. The solutions I plan on pursuing are:
1. create new keyed subreducer under comments with the shape: { [ `siteId-commentId` ]: true  }.  If a ket exists here that means we could not retrieve it.
2. add new selector for `getCommentById`.  If the component is able to get the comment successfully, but its for the wrong post, then request post normally.

File: client/blocks/comments/post-comment-list.jsx [hunk 5]
```diff
@@ -55,26 +71,81 @@ class PostCommentList extends React.Component {
...
-			( this.props.post.site_ID !== nextSiteId ||
-				this.props.post.ID !== nextPostId ||
-				this.props.commentsFilter !== nextCommentsFilter )
+		const nextInitialComment = nextProps.initialComment;
+
+		if ( this.shouldFetchInitialComment( nextProps ) && ! nextProps.initialComment ) {
+			this.props.requestComment( { siteId: nextSiteId, commentId: nextProps.startingCommentId } );
+			this.hasScrolledToComment = false;
+		} else if ( this.shouldFetchInitialPages( nextProps ) ) {
+			this.viewEarlierCommentsHandler();
```
Comment (blowery) [inline client/blocks/comments/post-comment-list.jsx]: doesn't this need to pass `nextProps` somehow to fetch the correct earlier and later pages?

File: client/blocks/comments/post-comment-list.jsx [hunk 6]
```diff
@@ -55,26 +71,81 @@ class PostCommentList extends React.Component {
...
+			nextProps.requestPostComments( {
+				siteId: nextSiteId,
+				postId: nextPostId,
+				status: nextCommentsFilter,
+			} );
+		}
+
+		if ( this.shouldScrollToComment( nextProps ) ) {
+			// defer so that the above/below comments have time to render
+			defer( () => this.scrollToComment(), 50 );
```
Comment (blowery) [inline client/blocks/comments/post-comment-list.jsx]: `defer` doesn't take a timeout value. it passes args on to the callback instead. do you mean to pass 50 or are you trying to wait 50ms?

File: client/blocks/comments/post-comment-list.jsx [hunk 7]
```diff
@@ -169,6 +240,13 @@ class PostCommentList extends React.Component {
 		);
 	};
 
+	scrollToComment = () => {
+		const comment = window.document.getElementsByName( window.location.hash.substring( 1 ) )[ 0 ];
+		comment.scrollIntoView();
```
Comment (blowery) [inline client/blocks/comments/post-comment-list.jsx]: this makes me a little nervous, but i'm not confident that `scrollIntoView` is a sync method. Are we guarenteed that the scroll is complete when the method returns?

File: client/blocks/comments/post-comment.jsx [hunk 8]
```diff
@@ -168,10 +168,20 @@ class PostComment extends Component {
 						: <Gravatar user={ comment.author } /> }
 
 					{ authorUrl
-						? <a href={ authorUrl } className="comments__comment-username" onClick={ this.handleAuthorClick }>
+						? <a
+							href={ authorUrl }
+							className="comments__comment-username"
+							onClick={ this.handleAuthorClick }
+							name={ `comment-${ this.props.commentId }` }
```
Comment (blowery) [inline client/blocks/comments/post-comment.jsx]: probably better to use `id` here rather than `name`. https://stackoverflow.com/a/484781

Comment (samouri) [inline client/blocks/comments/post-comment.jsx]: looks equivalent for this case, but i'll switch it over.  that was an interesting read, I had no idea how the spec worked before.

File: client/blocks/comments/post-comment-list.jsx (hunk 6 above)
Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: wow. good catch.  i was looking for `delay`.  I'll swap it out (although it was working anyway, so probably just need to wait for the callstack to clear).

File: client/blocks/comments/post-comment-list.jsx (hunk 5 above)
Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: I don't think so.  all of the comments are in redux and we use the oldest/newest comment in order to page.

**Comment by samouri** [comment]
We can _almost_ delete all of the `totalCommentsCount` code.  The only piece of functionality we still gain from it is to inform the CommentButton to increment/decrement when a user comments/deletes a comment.  (it happens differently within the comments list).

File: client/blocks/comments/post-comment-list.jsx (hunk 7 above)
Comment (samouri) [inline client/blocks/comments/post-comment-list.jsx]: good question, i have no answer.  mdn didn't have the answer either.
Do you know of a better guaranteed sync alternative?

Comment (blowery) [inline client/blocks/comments/post-comment-list.jsx]: i don't. like i said, just nervous. seems fine in testing.

Comment (blowery) [inline client/blocks/comments/post-comment-list.jsx]: https://readable-email.org/list/whatwg/topic/scrollintoview

File: client/state/comments/reducer.js [hunk 9]
```diff
@@ -190,15 +196,24 @@ export const totalCommentsCount = createReducer(
 export const errors = createReducer(
 	{},
 	{
-		[ COMMENTS_ERROR ]: ( state, action ) => ( {
-			...state,
-			[ `${ action.siteId }-${ action.commentId }` ]: { error: true },
-		} ),
+		[ COMMENTS_ERROR ]: ( state, action ) => {
```
Comment (blowery) [inline client/state/comments/reducer.js]: how would we unset this? say the network comes back up or it was a transient HTTP 500...

Comment (samouri) [inline client/state/comments/reducer.js]: Transient 500 should have been taken care of by automatic retries. 

As the reducer is written now, a hard refresh would be necessary.  But, note, this is only about errors loading a comment by id. We fallback to regular comment loading

**Comment by blowery** [comment]
load http://calypso.localhost:3000/read/blogs/16956016/posts/26154#comment-45555
notice the "load more" link at the bottom.
click it.
nothing appears to happen. 

can we detect when we've loaded the end of the list?

**Comment by samouri** [comment]
but something did happen! it loaded more comments...at the top!

**Comment by samouri** [comment]
In all seriousness though: its possible.  We could pre-emptively do a check in either direction and just keep the results hidden in state.  I think the behavior right now, while not perfect, is acceptable. I'd rather chase ux improvements as iterations after landing this PR.

If anything this could be a blocker for https://github.com/Automattic/wp-calypso/pull/14396

**Comment by blowery** [comment]
> can we detect when we've loaded the end of the list?

chatted with @samouri about this, we're going to accept it for this iteration. The coming comment tree project should help resolve it.

**Comment by yoavf** [comment]
@samouri the two instances of the "Load more comments..." string have different capitalization, which means duplicate translations efforts. Can we choose one?

**Comment by samouri** [comment]
@yoavf: that was my mistake.  Both should be "Load more comments".  I can fix that tomorrow

**Comment by samouri** [comment]
created the pr here: https://github.com/Automattic/wp-calypso/pull/16584

**Pull request merged**